Assets/Scripts/Encounters/ShopEncounter.cs

```diff
@@ -101,47 +102,71 @@ public override void build(List<Companion> companionList, EncounterConstantsSO c
         }
     }
 
-    private void generateShopEncounter() {
-        // Since we're generating the shop, lets clear the current shop
+    private void generateShopEncounter(List<Companion> companionList) {
         cardsInShop = new List<CardInShopWithPrice>();
         keepsakesInShop = new List<KeepsakeInShopWithPrice>();
 
-        generateCards();
+        generateCards(companionList);
         generateKeepsakes();
     }
 
-    private void generateCards() {
-        int numCardsToGenerate = shopData.cardLocations.Count;
-        int totalPercentage = shopData.getTotalCardPercentage();
-        int commonCardPercentage = shopData.commonCardPercentage;
-        int uncommonCardPercentage = shopData.uncommonCardPercentage;
-        int rareCardPercentage = shopData.rareCardPercentage;
+    private void generateCards(List<Companion> companionList) {
+        Debug.Log("Companion List count: " + companionList.Count);
+        //determine which companion to spawn a card from, remove them from the set
+        //move companion types to a hashSet
+        HashSet<CompanionTypeSO> companionTypes = new HashSet<CompanionTypeSO>();
+
+        foreach (Companion companion in companionList) {
+            companionTypes.Add(companion.companionType);
+        }
+
+        Debug.Log("Companion types: " + companionTypes.Count);
+
+        foreach (CompanionTypeSO companionType in companionTypes) {
+            //pick a random card based on random algorithm
+            CardPoolSO currentCardPool = companionType.cardPool;
+            int totalPercentage = currentCardPool.getTotalCardPercentage();
+            int commonCardPercentage = currentCardPool.commonCardPercentage;
+            int uncommonCardPercentage = currentCardPool.uncommonCardPercentage;
+            int rareCardPercentage = currentCardPool.rareCardPercentage;
+
 
-        for (int i = 0; i < numCardsToGenerate; i++) {
             // Determine what the card pool is for this single card being generated
-            List<CardType> cardPool = new List<CardType>();
+            SerializableHashSet<CardType> cardSet = new SerializableHashSet<CardType>();
             int cardPrice = 0;
             int randomNumber = Random.Range(0, totalPercentage); // min inclusive, max exclusive
             if (randomNumber < commonCardPercentage) {
-                cardPool = shopData.commonCards;
-                cardPrice = shopData.commonCardPrice;
+                cardSet = currentCardPool.commonCards;
+                cardPrice = currentCardPool.commonCardPrice;
```

Comment:
We talked it over a bit, and unless we know those details it made sense just to make it simple for now since just sticking the price to the card type SO didn't seem like the best answer either.